### ISSUE TITLE
LogSoftMax only implements log(softmax()), not a numerically more stable version.

### ISSUE THREAD

**Issue by ColaColin** [deeplearning4j/nd4j#2822]
#### Issue Description

LogSoftMax in pytorch and tensorflow are implemented using numerically more stable formulas that do not just call log(softmax()). In a deep ResNet I am seeing NaNs and unstable training because of this.

It seems nd4j does not have such a version of LogSoftMax. All they do is encapsule log(softmax()). This is not numerically stable in some cases and causes training to fail.

According to the [tensorflow documentation](https://www.tensorflow.org/api_docs/python/tf/nn/log_softmax) they use 
`logsoftmax = logits - log(reduce_sum(exp(logits), axis))`

I cannot find a formula for pytorch, only this [code](https://github.com/pytorch/pytorch/blob/4ad7fab16e814dc8077210a127317ec3ac620967/aten/src/THNN/generic/LogSoftMax.c), which seems to be going a similar path.

nd4j should implement the more stable version of LogSoftMax.

#### Version Information
<nd4j.version>1.0.0-alpha</nd4j.version>
<dl4j.version>1.0.0-alpha</dl4j.version>
